libraries/persistent_store/src/buffer.rs

```diff
@@ -0,0 +1,668 @@
+    /// starting at byte `index`, but actually only `value` was written. Word counters are
+    /// incremented only if their value would change and they would be completely written.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the maximum number of writes per word is reached.
+    fn incr_word_writes(&mut self, index: usize, value: &[u8], complete: &[u8]) {
+        let word_size = self.word_size();
+        for i in 0..value.len() / word_size {
+            let range = core::ops::Range {
+                start: i * word_size,
+                end: (i + 1) * word_size,
+            };
+            // Partial word writes do not count.
+            if value[range.clone()] != complete[range.clone()] {
+                continue;
+            }
+            // Words are written only if necessary.
+            if value[range.clone()] == self.storage[index..][range] {
+                continue;
+            }
+            let word = index / word_size + i;
+            assert!(self.word_writes[word] < self.max_word_writes());
+            self.word_writes[word] += 1;
+        }
+    }
+
+    /// Returns the storage range of an operation.
+    fn operation_range(
+        &self,
+        operation: &BufferOperation<impl Borrow<[u8]>>,
+    ) -> StorageResult<core::ops::Range<usize>> {
+        match *operation {
+            BufferOperation::Write { index, ref value } => index.range(value.borrow().len(), self),
+            BufferOperation::Erase { page } => {
+                StorageIndex { page, byte: 0 }.range(self.page_size(), self)
+            }
+        }
+    }
+}
+
+impl Storage for BufferStorage {
+    fn word_size(&self) -> usize {
+        self.options.word_size
+    }
+
+    fn page_size(&self) -> usize {
+        self.options.page_size
+    }
+
+    fn num_pages(&self) -> usize {
+        self.storage.len() / self.options.page_size
+    }
+
+    fn max_word_writes(&self) -> usize {
+        self.options.max_word_writes
+    }
+
+    fn max_page_erases(&self) -> usize {
+        self.options.max_page_erases
+    }
+
+    fn read_slice(&self, index: StorageIndex, length: usize) -> StorageResult<&[u8]> {
+        Ok(&self.storage[index.range(length, self)?])
+    }
+
+    fn write_slice(&mut self, index: StorageIndex, value: &[u8]) -> StorageResult<()> {
+        if !self.is_word_aligned(index.byte) || !self.is_word_aligned(value.len()) {
+            return Err(StorageError::NotAligned);
+        }
+        let operation = BufferOperation::Write { index, value };
+        let range = self.operation_range(&operation)?;
+        // Interrupt operation if armed and delay expired.
+        self.interruption.tick(&operation)?;
+        // Check and update counters.
+        self.incr_word_writes(range.start, value, value);
+        // Check strict write.
+        if self.options.strict_write {
+            for (byte, &val) in range.clone().zip(value.iter()) {
+                assert_eq!(self.storage[byte] & val, val);
```

Comment:
Not sure what "commands" and "rather panic" mean. But here is more context about strict writes. It is (almost) always enabled because that's how flashs work. So checking it permits to catch bugs in the store implementation. The only case where it's set to false is when we fuzz starting from an initial corrupted state. In that case we only check for crashes because the store can't work properly, so in particular we disable this check. Note that writing to the flash in a "non-strict" way is permitted, it just doesn't write the bits that are flipped in the wrong direction.

Reply:
Ah okay, so that is just a feature simulating flash storage. And you can disable it for testing / fuzzing.